### DISCUSSION TITLE
How to type functions.

### DISCUSSION THREAD

**Discussion by ff6347** [technologiestiftung/giessdenkiez-de#297] [General]
I'd like to ask a question/start a discussion about typing functions.
I saw that you (@vogelino) tend to write signatures like this:

```ts
const setVisibleMapLayer = (
  type: StoreProps['visibleMapLayer']
): { visibleMapLayer: StoreProps['visibleMapLayer'] } => {
  return {
    visibleMapLayer: type,
  };
};
```
https://github.com/technologiestiftung/giessdenkiez-de/blob/612d5f0cc78be4433f9d0387bdf8c62a6b6c92a9/src/state/Actions.tsx#L13

Where I like to write them like this:

```ts
const setVisibleMapLayer: (
  type: StoreProps['visibleMapLayer']
) => { visibleMapLayer: StoreProps['visibleMapLayer'] } = (type) => {
  return {
    visibleMapLayer: type,
  };
};
```

Why do I prefer the one over the other?

If I would have to share this function signature with another function I just could do extract it by copying it into a type and I'm done.

```ts
type SetterFunc =  (
  type: StoreProps['visibleMapLayer']
) => { visibleMapLayer: StoreProps['visibleMapLayer'] };

const setVisibleMapLayer: SetterFunc= (type) => {
  return {
    visibleMapLayer: type,
  };
};
```

In your way of typing it the description of the types and the implementation are tightly coupled. Any reason why you write it your way?

---

Another thing I do often is use the `function` keyword on top level functions.

```ts
function(a: number): number {
return a;
}
```

So my rule is to only use `const thing = () …` if I'm in a closure and use `function` in the global scope. Any reason why you don't use `function` at all?

@lucasoeth @dnsos I'd also like to hear your opinion about `this`.

---

Lets assume that all the things said in #293 are below.

**Comment by ff6347** [comment]
…
@vogelino 

> Common, could we not talk about your way/ my way?

Yes. Sorry. 

> An can you elaborate on why do you think it is easier to digest?

I think if the reader is used to that kind of signature they can scan from `:` until `=` to see the definition. Since I'm used to it I had problems reading the tighter coupled way at first sight.

**Reply by vogelino** [reply]
I agree, it is a question of habit. I'm not too religious about it.

**Reply by ff6347** [reply]
me neither. We currently have more of tightly coupled signatures. Let's keep it that way.
